### PR TITLE
Optimizations in `symbolic.py`

### PR DESCRIPTION
**Pull request by rayanht**
# Summary

This diff modifies the `symbolic.py` file mainly to not rely on calls to the expensive `isinstance` by leveraging inheritance to dispatch the correct routines instead.

Changelog:
1. Get rid of `create_opnode` because it had a massive if-else ladder of checking the class type, replace by computing minmax bounds at init time of `OpNode`
2. Move logic for `__mul__`, `__floordiv__`, and `__mod__` inside each opcode
3. Rewrite the `Variable.sum` simplification routine to avoid having to order the list of `MulNodes` before grouping them

# Test Plan

## Correctness

Local test suite is passing. Wait for CI

## Performance

`test_net_speed.py` benchmark:

### Before

```
test/test_net_speed.py::TestConvSpeed::test_mnist torch forward pass:  19.031 ms
torch backward pass: 19.433 ms
         475052 function calls (464120 primitive calls) in 280.306 seconds

   Ordered by: internal time
   List reduced from 593 to 119 due to restriction <0.2>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   132929   20.959    0.000   20.959    0.000 {built-in method builtins.isinstance}
11513/11458   14.423    0.001   34.881    0.003 symbolic.py:28(__mul__)
     8384    9.673    0.001   15.625    0.002 symbolic.py:170(create_opnode)
9207/6134    7.457    0.001   13.082    0.002 symbolic.py:13(render)
     1142    6.776    0.006   78.155    0.068 symbolic.py:109(sum)
        5    6.689    1.338    6.921    1.384 ops_gpu.py:44(_copyin)
    13180    5.426    0.000    7.322    0.001 symbolic.py:118(<lambda>)
      791    5.350    0.007   45.426    0.057 shapetracker.py:185(reshape)
       16    5.255    0.328    5.255    0.328 {built-in method io.open}
   910/20    4.914    0.005  175.367    8.768 lazy.py:107(realize)
     1845    4.770    0.003    7.997    0.004 shapetracker.py:33(__init__)
     1142    4.062    0.004   31.675    0.028 symbolic.py:126(<listcomp>)
1200/1035    3.753    0.003   79.001    0.076 lazy.py:187(movement_op)
1933/1655    3.464    0.002    7.902    0.005 symbolic.py:43(__floordiv__)
     9006    3.377    0.000    3.944    0.000 symbolic.py:156(create_node)
     7998    3.338    0.000    4.493    0.001 symbolic.py:123(<lambda>)
     7998    3.330    0.000   11.514    0.001 symbolic.py:126(<genexpr>)
     5893    3.289    0.001   10.449    0.002 symbolic.py:20(__eq__)
     2284    3.258    0.001    9.107    0.004 helpers.py:15(<listcomp>)
    11150    3.227    0.000    3.227    0.000 shapetracker.py:34(<genexpr>)
  1680/80    3.195    0.002    9.183    0.115 dataclasses.py:1280(_asdict_inner)
      520    2.986    0.006   41.359    0.080 shapetracker.py:61(expr_node)
17256/15872    2.974    0.000    3.466    0.000 {built-in method builtins.len}

forward pass:  16.695 ms, 0.88x off baseline 19.031 ms
backward pass: 40.735 ms, 2.10x off baseline 19.433 ms
```

### After

```
test/test_net_speed.py::TestConvSpeed::test_mnist torch forward pass:  19.160 ms
torch backward pass: 20.129 ms
         361819 function calls (351235 primitive calls) in 238.417 seconds

   Ordered by: internal time
   List reduced from 592 to 118 due to restriction <0.2>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1142   12.107    0.011   42.137    0.037 symbolic.py:59(sum)
        5   10.644    2.129   10.822    2.164 ops_gpu.py:44(_copyin)
    55789    8.929    0.000    8.929    0.000 {built-in method builtins.isinstance}
    11876    7.517    0.001   18.580    0.002 symbolic.py:30(__mul__)
8036/4978    6.466    0.001   11.452    0.002 symbolic.py:15(render)
        4    6.166    1.541    6.166    1.541 {built-in method posix.unlink}
    10086    5.767    0.001    9.687    0.001 symbolic.py:120(__init__)
       16    5.684    0.355    5.684    0.355 {built-in method io.open}
      791    5.028    0.006   33.577    0.042 shapetracker.py:185(reshape)
     1845    4.653    0.003    7.756    0.004 shapetracker.py:33(__init__)
   910/20    4.559    0.005  143.841    7.192 lazy.py:107(realize)
        4    4.387    1.097    4.387    1.097 {built-in method posix.open}
     9488    3.644    0.000    3.644    0.000 symbolic.py:145(get_bounds)
1200/1035    3.622    0.003   66.761    0.065 lazy.py:187(movement_op)
     9470    3.112    0.000    3.647    0.000 symbolic.py:114(create_node)
    11150    3.103    0.000    3.103    0.000 shapetracker.py:34(<genexpr>)
     1142    3.086    0.003   12.706    0.011 symbolic.py:84(<listcomp>)
  1680/80    3.020    0.002    8.799    0.110 dataclasses.py:1280(_asdict_inner)
     1475    2.975    0.002   49.009    0.033 shapetracker.py:230(movement_op)
     6610    2.927    0.000    4.568    0.001 {built-in method builtins.sum}
     
forward pass:  15.501 ms, 0.81x off baseline 19.160 ms
backward pass: 32.857 ms, 1.63x off baseline 20.129 ms
```

We save roughly `80,000` calls to `isinstance` from `132,929` to `55,789` **(-58%)**. 

Forward pass on average **9%** faster
Backward pass on average **18%** faster

# Future iterations

There's still `isinstance` in a _lot_ of places in the codebase, might be worth trying to remove as many as possible. Would be down to work on that.

I also had some optimizations on `shapetracker.py` but it's probably better I submit them in a separate PR

### CODE REVIEW

File: test/test_net_speed.py [hunk 1]
```diff
@@ -18,6 +18,8 @@ def stop_profile(pr, sort='cumtime'):
   ps.strip_dirs()
   ps.sort_stats(sort)
   ps.print_stats(0.2)
+  filename = 'profile.prof'
```
Comment (rayanht) [inline test/test_net_speed.py:21]: TODO: remove when ready to propose PR

File: tinygrad/shape/symbolic.py [hunk 2]
```diff
@@ -88,47 +47,43 @@ def __floordiv__(self, b:int, factoring_allowed=True):
...
-    new_nodes = [k * sum(x.b for x in g) for k, g in itertools.groupby(mul_nodes, key=lambda x: x.a)]
-    nodes = [x if not isinstance(x, MulNode) or x.b != 1 else x.a for x in new_nodes]
-
-    # filter 0s
-    nodes = [x for x in nodes if x.min != 0 or x.max != 0]
-    return create_rednode(SumNode, nodes) if len(nodes) > 1 else (nodes[0] if len(nodes) == 1 else NumNode(0))
+    for node in nodes:
+      if isinstance(node, NumNode):
+        return node
+      if isinstance(node, MulNode):
```
Comment (rayanht) [inline tinygrad/shape/symbolic.py:62]: TODO: handle this case and clean up comments

**Comment by rayanht** [comment]
Hey @geohot could you approve the workflow even though this is still in draft? There's some tests I can't run on my local machine

**Comment by geohot** [comment]
Approved, but unfortunately I have to reapprove it every time.

Re: optimizations, want to focus on one thing at a time? I see you rewrote a lot of symbolic, are you sure the tests are good enough? There's a lot of subtlety there.

**Comment by rayanht** [comment]
@geohot I figured in the meantime how to get everything running locally so no worries about the CI.

re: the symbolic stuff, it looks like i changed a lot of things but 90% of it is preserving the exact same logic that was there but moving it around to leverage inheritance instead of calling `isinstance` everywhere because those are expensive af

**Comment by geohot** [comment]
Yea, smart idea. `isinstance`is terrible.

**Comment by rayanht** [comment]
# Summary

This diff modifies the `symbolic.py` file to not rely on calls to the expensive `isinstance` by leveraging inheritance to dispatch the correct routines instead.

# Test Plan

## Correctness

Local test suite is passing. Wait for CI

## Performance

`test_net_speed.py` benchmark:

### Before

```
test/test_net_speed.py::TestConvSpeed::test_mnist torch forward pass:  19.002 ms
torch backward pass: 34.680 ms
         108427 function calls (107387 primitive calls) in 58.889 seconds

   Ordered by: internal time
   List reduced from 231 to 46 due to restriction <0.2>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    23100    3.645    0.000    3.645    0.000 {built-in method builtins.isinstance}
      335    2.440    0.007   24.916    0.074 shapetracker.py:185(reshape)
      890    2.139    0.002    3.565    0.004 shapetracker.py:33(__init__)
  460/410    1.676    0.004   41.482    0.101 lazy.py:187(movement_op)
1535/1505    1.475    0.001    3.530    0.002 symbolic.py:28(__mul__)
      685    1.459    0.002   33.434    0.049 shapetracker.py:230(movement_op)
     5405    1.426    0.000    1.426    0.000 shapetracker.py:34(<genexpr>)
      895    1.351    0.002    4.485    0.005 {built-in method builtins.all}
      215    1.276    0.006    8.174    0.038 symbolic.py:109(sum)
      410    1.143    0.003    2.744    0.007 lazy.py:88(__init__)
      955    1.118    0.001    1.916    0.002 symbolic.py:170(create_opnode)
      445    1.068    0.002   49.496    0.111 tensor.py:19(apply)
  665/585    1.046    0.002    2.538    0.004 symbolic.py:43(__floordiv__)
      465    1.041    0.002    4.279    0.009 tensor.py:36(__init__)
      215    1.011    0.005   14.750    0.069 shapetracker.py:61(expr_node)
6065/5705    1.001    0.000    1.124    0.000 {built-in method builtins.len}
      445    0.976    0.002    7.012    0.016 lazy.py:70(create_lazybuffer)
     4280    0.942    0.000    0.942    0.000 shapetracker.py:115(shape)
     2080    0.879    0.000    1.135    0.001 shapetracker.py:187(<genexpr>)
     1925    0.825    0.000    1.174    0.001 enum.py:1229(__hash__)
   420/20    0.766    0.002    2.162    0.108 dataclasses.py:1277(_asdict_inner)
      445    0.728    0.002    1.325    0.003 tensor.py:11(__init__)
      110    0.659    0.006    3.097    0.028 shapetracker.py:214(permute)
     1600    0.645    0.000    0.870    0.001 symbolic.py:118(<lambda>)
      780    0.645    0.001    1.007    0.001 shapetracker.py:215(<genexpr>)
       35    0.640    0.018   17.152    0.490 shapetracker.py:86(merge_views)
      200    0.636    0.003   24.930    0.125 tensor.py:200(reshape)
  425/415    0.564    0.001    1.330    0.003 lazy.py:57(get_weakop)
      585    0.548    0.001    1.263    0.002 symbolic.py:88(__mod__)
     1255    0.515    0.000    0.824    0.001 {built-in method builtins.any}
      825    0.511    0.001    0.663    0.001 shapetracker.py:107(__init__)
     1045    0.500    0.000    0.611    0.001 shapetracker.py:180(<genexpr>)
      380    0.491    0.001    1.178    0.003 weakref.py:164(__setitem__)
     1300    0.475    0.000    0.616    0.000 shapetracker.py:194(<genexpr>)
      970    0.465    0.000    0.647    0.001 shapetracker.py:112(contiguous)
  565/545    0.449    0.001    0.746    0.001 symbolic.py:13(render)
       20    0.437    0.022   24.347    1.217 tensor.py:331(_pool)
      430    0.433    0.001    1.028    0.002 helpers.py:15(<listcomp>)
      955    0.428    0.000    0.546    0.001 symbolic.py:156(create_node)
      430    0.418    0.001    2.474    0.006 helpers.py:15(partition)
      320    0.416    0.001    0.639    0.002 copy.py:128(deepcopy)
  920/890    0.413    0.000    0.518    0.001 lazy.py:57(<genexpr>)
      870    0.405    0.000    0.586    0.001 <string>:1(<lambda>)
      175    0.403    0.002    2.019    0.012 shapetracker.py:179(expand)
  420/410    0.396    0.001    0.847    0.002 ops.py:28(get_buffers)
      215    0.391    0.002    2.438    0.011 symbolic.py:126(<listcomp>)


forward pass:  11.930 ms, 0.63x off baseline 19.002 ms
backward pass: 0.141 ms, 0.00x off baseline 34.680 ms
```

### After

```
test/test_net_speed.py::TestConvSpeed::test_mnist torch forward pass:  20.084 ms
torch backward pass: 23.449 ms
         93792 function calls (92862 primitive calls) in 53.364 seconds

   Ordered by: internal time
   List reduced from 225 to 45 due to restriction <0.2>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      335    2.460    0.007   18.745    0.056 shapetracker.py:186(reshape)
      890    2.128    0.002    3.540    0.004 shapetracker.py:34(__init__)
    12045    1.942    0.000    1.942    0.000 {built-in method builtins.isinstance}
      215    1.889    0.009    5.075    0.024 symbolic.py:59(sum)
  460/410    1.654    0.004   35.413    0.086 lazy.py:187(movement_op)
      685    1.533    0.002   27.355    0.040 shapetracker.py:231(movement_op)
     5405    1.412    0.000    1.412    0.000 shapetracker.py:35(<genexpr>)
      895    1.130    0.001    4.009    0.004 {built-in method builtins.all}
      410    1.119    0.003    2.746    0.007 lazy.py:88(__init__)
      445    1.085    0.002   43.460    0.098 tensor.py:19(apply)
      465    1.071    0.002    4.323    0.009 tensor.py:36(__init__)
6065/5705    1.056    0.000    1.186    0.000 {built-in method builtins.len}
     4280    1.004    0.000    1.004    0.000 shapetracker.py:116(shape)
      215    0.991    0.005    9.283    0.043 shapetracker.py:62(expr_node)
     1650    0.987    0.001    2.344    0.001 symbolic.py:30(__mul__)
      445    0.966    0.002    7.022    0.016 lazy.py:70(create_lazybuffer)
     1925    0.848    0.000    1.176    0.001 enum.py:1229(__hash__)
     1345    0.815    0.001    1.365    0.001 symbolic.py:120(__init__)
   420/20    0.759    0.002    2.186    0.109 dataclasses.py:1277(_asdict_inner)
      445    0.751    0.002    1.329    0.003 tensor.py:11(__init__)
       35    0.689    0.020   11.409    0.326 shapetracker.py:87(merge_views)
      200    0.675    0.003   20.623    0.103 tensor.py:200(reshape)
      780    0.674    0.001    1.048    0.001 shapetracker.py:216(<genexpr>)
      110    0.658    0.006    3.097    0.028 shapetracker.py:215(permute)
     1430    0.612    0.000    0.774    0.001 shapetracker.py:188(<genexpr>)
  670/650    0.599    0.001    0.910    0.001 symbolic.py:15(render)
  425/415    0.564    0.001    1.339    0.003 lazy.py:57(get_weakop)
      825    0.538    0.001    0.672    0.001 shapetracker.py:108(__init__)
     1045    0.530    0.001    0.660    0.001 shapetracker.py:181(<genexpr>)
      380    0.505    0.001    1.144    0.003 weakref.py:164(__setitem__)
      970    0.504    0.001    0.674    0.001 shapetracker.py:113(contiguous)
     1040    0.487    0.000    0.785    0.001 {built-in method builtins.any}
     1300    0.454    0.000    0.592    0.000 shapetracker.py:195(<genexpr>)
     1125    0.449    0.000    0.449    0.000 symbolic.py:145(get_bounds)
      320    0.441    0.001    0.647    0.002 copy.py:128(deepcopy)
     1115    0.434    0.000    0.550    0.000 symbolic.py:114(create_node)
       20    0.423    0.021   20.026    1.001 tensor.py:331(_pool)
  420/410    0.420    0.001    0.871    0.002 ops.py:28(get_buffers)
      175    0.411    0.002    2.016    0.012 shapetracker.py:180(expand)
  920/890    0.409    0.000    0.521    0.001 lazy.py:57(<genexpr>)
      870    0.407    0.000    0.599    0.001 <string>:1(<lambda>)
     1560    0.397    0.000    0.397    0.000 shapetracker.py:193(<genexpr>)
     1310    0.393    0.000    0.393    0.000 symbolic.py:111(__init__)
     2460    0.388    0.000    0.388    0.000 {method 'append' of 'list' objects}
       60    0.376    0.006    4.316    0.072 tensor.py:213(slice)


forward pass:  10.745 ms, 0.54x off baseline 20.084 ms
backward pass: 0.142 ms, 0.01x off baseline 23.449 ms
```

We roughly halve the number of calls to `isinstance` from 23100 to 12045. Performance gain on this benchmark is on average 10-20%

**Comment by geohot** [comment]
Hmm, I understand, smart. Treat the specializations as overrides instead of checks in the base class.

Did you run the fuzzer in `fuzz_symbolic.py`?

**Comment by rayanht** [comment]
Didn't realize there was a fuzzer, will run it for a bit (would 10 minutes be enough?) and report back on whether or not it crashes

**Comment by geohot** [comment]
Oh yea, if it doesn't hit in a minute or two it probably won't. I found a few bugs with it.

**Comment by rayanht** [comment]
aite yeah fuzzer ran fine, will fix the linter CI and look into why OpenCL isn't happy

**Comment by geohot** [comment]
Oh hehe, the OpenCL one is real.

**Comment by geohot** [comment]
Btw, that test_net_speed was so old it didn't mark the gradients as requiring grad and it didn't realize the tensors. Fixed.

https://github.com/geohot/tinygrad/commit/ea3194f68e69959d7b90556e2219974a898982c1#diff-3161078a1af15bdd0133ddc73d9c69214dda252d2381027ff2b78d2294e005e4

**Comment by rayanht** [comment]
^no need to run the CI for this last push, was just pulling changes from trunk, still investigating the OpenCL thing.

The perf gain is less significant with the corrected test case but still not too bad, updated the summary 

We save roughly 15,000 calls to isinstance from 90,030 to 75,170. Performance gain is on average 5% in the forward pass and 4% in the backward pass.

**Comment by geohot** [comment]
Oh, you don't have a GPU? `pip install pyopencl` if you do, it's default with tinygrad now. If there's no GPU, it uses the terrible numpy backend which is dominating that profile.

**Comment by rayanht** [comment]
Working from my Mac, I do have a GPU on my desktop let me switch to that then

**Comment by geohot** [comment]
Mac GPU is well supported.

**Comment by rayanht** [comment]
Alright so with OpenCL I get the following results:

Save roughly `80,000` calls to isinstance from `132,929` to `55,789` **(-58%)**.

Forward pass on average **9%** faster
Backward pass on average **18%** faster

Updated summary + still looking into the sneaky test case

**Comment by rayanht** [comment]
Well this was fun, took me longer to figure out this single test case than the rest of the PR combined. Anyway it passes locally now and performance numbers are unchanged

**Comment by geohot** [comment]
Sweet! Can you add an explicit test for it to `test_symbolic`? Got lucky that it was caught by that random one.

**Comment by rayanht** [comment]
Ok spent some time hunting exactly where things go wrong in the OpenCL test and came up with minimal example `def test_sum_num_hoisted_and_factors_cancel_out`

**Comment by geohot** [comment]
Amazing!

**Comment by geohot** [comment]
On a tinybox

From:
```forward pass:  25.778 ms, 1.74x off baseline 14.781 ms
backward pass: 62.059 ms, 1.24x off baseline 49.959 ms```

To:
```forward pass:  23.422 ms, 1.55x off baseline 15.157 ms
backward pass: 48.914 ms, 1.00x off baseline 49.085 ms```

**Comment by rayanht** [comment]
Sick!

**Pull request merged**